File: conformance/tests/token-nft/token-nft.runar.ts
import { StatefulSmartContract, assert, checkSig } from 'runar-lang';

class SimpleNFT extends StatefulSmartContract {
  owner: PubKey;
  readonly tokenId: ByteString;
  readonly metadata: ByteString;

  constructor(owner: PubKey, tokenId: ByteString, metadata: ByteString) {
    super(owner, tokenId, metadata);
    this.owner = owner;
    this.tokenId = tokenId;
    this.metadata = metadata;
  }

  public transfer(sig: Sig, newOwner: PubKey, outputSatoshis: bigint) {
    assert(checkSig(sig, this.owner));
    assert(outputSatoshis >= 1n);
    this.addOutput(outputSatoshis, newOwner);
  }

  public burn(sig: Sig) {
    assert(checkSig(sig, this.owner));
  }
}

File: conformance/tests/token-nft/token-nft.runar.rs
use runar::prelude::*;

#[runar::contract]
struct SimpleNFT {
    owner: PubKey,
    #[readonly]
    token_id: ByteString,
    #[readonly]
    metadata: ByteString,
}

#[runar::methods(SimpleNFT)]
impl SimpleNFT {
    #[public]
    fn transfer(&mut self, sig: Sig, new_owner: PubKey, output_satoshis: Bigint) {
        assert!(check_sig(sig, self.owner));
        assert!(output_satoshis >= 1);
        self.add_output(output_satoshis, new_owner);
    }

    #[public]
    fn burn(&mut self, sig: Sig) {
        assert!(check_sig(sig, self.owner));
    }
}
